### DISCUSSION TITLE
self hosted installation

### DISCUSSION THREAD

**Discussion by CircleCode** [knowledgefutures/pubpub#1470] [Support]
Hi, it seems installation (or run) instructions are not enough to start a self-hosted instance, at least for testing.
Do you have detailed how-to for local installation?

Following are the notes I took:
- node 14 is not supported, I had to stick to node 12
- npm start triggers a lot of errors about missing `config.js` file ⇒ I had to copy the `config.sample.js` file, but then I do not know which values are mandatory and which are not
  - I suppose a postgresql connection is required: is it enough to simply create a db (and what is the form of `DATABASE_URL`)?
- after digging a bit in the code, I see that it should be able to run without DATABASE_URL (per https://github.com/pubpub/pubpub/blob/f913b01c1cdef9a35189e6ba41446bbd9faa9ad2/stubstub/global/setup.js#L13), but then it seems per https://github.com/pubpub/pubpub/blob/f913b01c1cdef9a35189e6ba41446bbd9faa9ad2/stubstub/testDatabase.js#L60-L64 that it would require both write permission for `/usr/local/var/pubpub-postgres-test` and an `initdb` script

**Comment by gabestein** [comment]
Hello! Thanks for the interest in running PubPub.

We're still not at a place where we're really supporting self-installs, as that's just not our growth focus at the moment. We should make that clearer in the README.

You're still welcome to give it a shot, of course, we just aren't able to provide much support at the moment. You can see more about required vs. optional dependencies [here](https://github.com/pubpub/pubpub/issues/1125).

Regarding the database, I believe what you've picked up on is our test suite. We currently use an online development db, but are discussing moving to local databases soon, which would make it easier to self-run.

**Reply by CircleCode** [reply]
Thanks, and sorry for not founding #1125 before.
As I understand it, it is currently not possible to install pubpub without using the proprietary service firebase, right?
For S3, open source compliant implementations exist, and for other components, I did not search.

Is this situation susceptible to evolve in a near future?

Thanks

**Reply by gabestein** [reply]
No worries! Unfortunately, yes, Firebase is a pretty deep proprietary dependency at the moment, and unlike S3, Algolia, etc., has no compatible open-source alternatives that we're aware of.

We do have half an eye on [Supabase](https://supabase.io/), and would love to move to something more like it at some point. But it's not quite there yet (in particular, we need functions), the migration would be quite complex, and we'd probably still rely on their hosted service — though, of course, it would make it much easier for folks who want to implement. So, not in a near future, no.

**Reply by gabestein** [reply]
I wanted to pop in and add some context here. It won't be easy, but we have plan to remove proprietary dependencies and offer better support for self-hosting. We're committed to getting there, and it's baked into our sustainability models. https://notes.knowledgefutures.org/pub/kkcevu38/release/3

**Reply by almereyda** [reply]
Thank you @gabestein for the extended context on this long-running conversation.

To continue to use this ticket as a technical frame for the progress on this, I'd like to summarise what we've just read:

- There is a need to consider technical development together with the financial capabilities of the KFN organisation, leading to a sustainable business model for the not-for-profit. This involves communication with the community and partners.

Taken this for granted, the technical development is described as:

- Replacing the Heroku build and release process with a "cloud native", containerised version.
- Replacing Amazon S3 with compatible APIs (Minio, Ceph RADOS, ...)
- Replacing Firebase with a similar real-time API (Supabase et al.)
- Replacing Heap with a locally runnable analytics suite (PostHog et al.)
- Replacing Algolia Seach with a comparative search index (Elasticsearch et al.)

To extend on the last case of search, and since no preliminary issue/discussion/projectcard was discoverable on that subject, I'd like to add that https://alternativeto.net/software/algolia/?license=opensource also knows some Open Source projects, which claim to be Algolia drop-in replacements. Had those been considered before?

- https://github.com/meilisearch/meilisearch (Rust)
- https://github.com/typesense/typesense (C++)

Also the relicenced Elasticsearch-fork OpenSearch seems to receive frequent updates:

- https://github.com/opensearch-project/OpenSearch

**Reply by gabestein** [reply]
Thanks, we'll take a look — but I want to note again that one of the factors here is sustainability. Open-source alternatives to closed software disappear constantly and then have to be swapped out. Due to this, we're somewhat uncomfortable adopting early versions of alternatives, particularly those run by for-profit companies with no wide community adoption that have yet to find revenue traction (Meilisearch, Typesense, and frankly even Supabase and PostHog all fall into this category). OpenSearch is a little too bare-bones for our use-case, and its self-styled "community-driven" label is kind of hilarious given that its chief maintainer is Amazon (though it does give us more confidence in its long-term sustainability).

All to say: just because alternatives exist do not mean they are the right choice for us.

**Reply by almereyda** [reply]
Thank you for pointing that out. It is very compelling to watch how careful and long-term oriented you and your organisation approach the space.

**Comment by cfcohen** [comment]
I asked in discussion topic, but I'll ask here as well.   Has anyone actually been able to run "npm start" and connect to "localhost:9876" as the README suggests?

**Reply by eastofesten** [reply]
Hello! Would love to know specifically where you're running into issues on this step. We don't offer support for self-installs for 
PubPub Legacy, though we are developing strong documentation for self-hosting PubPub Platform.

**Reply by cfcohen** [reply]
I switched to PubPubPlatform and got further.  I'm now blocked on user creation.  I'll ask about that in the appropriate repo.  Thanks for responding.

**Comment by emizentechae** [comment]
Looking to launch your own fantasy sports platform? Our [fantasy sports app development company](https://emizentech.ae/fantasy-app-development.html) builds engaging, feature-rich, and scalable apps for cricket, football, basketball, and more. From real-time analytics to secure payment integration, we deliver customized solutions that keep users hooked. Get ahead in the booming fantasy sports market with our expert development team.
